### ISSUE TITLE
The results from the converted model doesn't match the ONNX original version.

### ISSUE THREAD

**Issue by PetiteFleurPF** [PINTO0309/onnx2tf#656]
### Issue Type

Others

### OS

Linux

### onnx2tf version number

1.23.0

### onnx version number

1.15.0

### onnxruntime version number

1.17.1

### onnxsim (onnx_simplifier) version number

0.4.33

### tensorflow version number

2.16.1

### Download URL for ONNX without post-processing

https://we.tl/t-Z6RrkZRZxf (TFLITE version)
https://we.tl/t-PH8hLCyGu9 (ONNX version)

### Parameter Replacement JSON

```json
*
```

### Description

1. Product development - impact: could be improved inference time to obtain TFLITE version.

2. The results from the TFLITE version of our model look like random bounding box coordinates (in size and location) when you plot them.
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/379a5807-e5d6-47d5-b183-87906396de71)

3. I investigated the differences between both results - they appeared before the post-processing part. So I tried to observe the differences, one of which was the observation of node deletion during conversion.
During the conversion process, some nodes disappeared: 
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/ee7b85e5-2ae1-4b62-b026-bde7e6a1cd97)
I tested every option, even the one that stopped the simplification and wasn't the solution. It is very surprising because these nodes existed during a moment because they appeared during process steps and validation:
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/3be25e40-c312-41fb-9320-af6f89204d7c)
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/a9477721-f874-4081-b20c-a2261bc49964)

 4. To succeed in the conversion of the converted ONNX model.
 
 5. You don't need more.

 PS: the previous problem was solved by only deleting the NMS part and not everything into the post-processing process but you helped us a lot to underline the correct way to do things. So thank you.

**Comment by PINTO0309** [comment]
I don't see the correlation between the attached ONNX file and the image you posted.

![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/7d458b0e-a134-4332-b5d3-ec3a55c14c69)

**Comment by PetiteFleurPF** [comment]
I can't send you the final model, the one I shared with you is similar and reproduces the same error. All the images I've pasted are from the final template.

**Comment by PetiteFleurPF** [comment]
For example, after use the model that I shared with you with the pipeline 
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/289dbcb4-ec4b-4c91-b900-78e0c8005c82)
I obtained that: 
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/7ee51628-48b4-4ffa-8bd5-a08522e4f277)
and that: 
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/d44d2a42-8667-4aa9-b09c-7f13b31d52f9)

**Comment by PINTO0309** [comment]
I still don't understand what your concern is. The ONNX DepthwiseConv2d -> Clip -> Conv2d -> Reshape -> Transpose -> Reshape section you illustrate is a rather redundant and useless combination of multiple OPs. I have no idea what the problem is with changing the shape of [1,24,1,1] to [1,6,4]. Your ONNX file is too much useless processing.



- ONNX
  ![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/00cf06c5-48fd-427e-b2a5-3d95a76ecb57)
- TFLite
  ![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/05fc4a82-c1a7-4153-bd84-2b946677f583)

**Comment by PetiteFleurPF** [comment]
OK - I'm in the process of removing the post-processing from the model so that we can start from the same observation, which is that the bounding boxes produced by the model are purely random - this can be seen in the various sizes and locations. In short, the converted version seems to have lost all the information obtained during training. I'll write again when the model is uploaded

**Comment by PetiteFleurPF** [comment]
And I try to understand why. That is why I made a hypothesis about the modification of the nodes

**Comment by PINTO0309** [comment]
![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/c6b7ba48-031d-4736-a940-8468fe2c1d41)

**Comment by PINTO0309** [comment]
```
onnx2tf -i model.onnx \
-onimc /head/regression_head/Concat_12_output_0 /Softmax_output_0 \
-cotof
```
![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/87a78a44-19ee-4f63-961d-b4aaea84d371)

![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/0214f8c3-38d5-425c-a718-3312c9b2f324)

**Comment by PINTO0309** [comment]
The myriad uses of `NonZero` in post-processing severely inhibit the normal model transformation behavior of onnx2tf because the output is non-deterministic.

![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/48831ac9-63bf-4394-94ea-55018c22b6c4)

**Comment by PINTO0309** [comment]
If you want to implement NMS, bounding box filtering by NonZero + TopK + If is quite redundant and quite wasteful to use for inference in TFLite

**Comment by PetiteFleurPF** [comment]
I still not updated the model ^^ Don't focus on the post-processing - the problem, my problem is not coming from there :) you will see (I am deleting them now (the post-process) to recreate a model without it to let you do a correct investigation) give me 5minutes. 

> ```
> onnx2tf -i model.onnx \
> -onimc /head/regression_head/Concat_12_output_0 /Softmax_output_0 \
> -cotof
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/33194443/345817014-87a78a44-19ee-4f63-961d-b4aaea84d371.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwOTM0NTAsIm5iZiI6MTcyMDA5MzE1MCwicGF0aCI6Ii8zMzE5NDQ0My8zNDU4MTcwMTQtODdhNzhhNDQtMTllZS00ZjYzLTk2MWQtYjRhYWVhODRkMzcxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDExMzkxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzNjI0MTBkNTQxNjQ0OTc5NTM1NDdiYzhiN2RkMzBkMTkyYjJmYjBmNTEyZTU3MmE4YjliMGUwMTRhMzdkMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.38Hmqqo07oJNqEF43DlReyBqsvgZ5_TXBQSvKuK3iv4)
> 
> ![image](https://private-user-images.githubusercontent.com/33194443/345817137-0214f8c3-38d5-425c-a718-3312c9b2f324.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwOTM0NTAsIm5iZiI6MTcyMDA5MzE1MCwicGF0aCI6Ii8zMzE5NDQ0My8zNDU4MTcxMzctMDIxNGY4YzMtMzhkNS00MjVjLWE3MTgtMzMxMmM5YjJmMzI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDExMzkxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2M2M5MjM2OWEzMzJkOTE5Y2VmMjYzOWQyZjk2NTc2OGRhZmZmYTgxMjBlMmMzNTZiOWYwYmVkOWViZGY4NzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XREPbiF495q5jWsdRX4VkeCwP-8aog97YfLl3ITSO3A)
 and here you reproduced what I said :) and after that, it's the post-processing, but at this moment, the output produced by the node "concat" is incorrect.

**Comment by PetiteFleurPF** [comment]
> ```
> onnx2tf -i model.onnx \
> -onimc /head/regression_head/Concat_12_output_0 /Softmax_output_0 \
> -cotof
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/33194443/345817014-87a78a44-19ee-4f63-961d-b4aaea84d371.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwOTM0NTAsIm5iZiI6MTcyMDA5MzE1MCwicGF0aCI6Ii8zMzE5NDQ0My8zNDU4MTcwMTQtODdhNzhhNDQtMTllZS00ZjYzLTk2MWQtYjRhYWVhODRkMzcxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDExMzkxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzNjI0MTBkNTQxNjQ0OTc5NTM1NDdiYzhiN2RkMzBkMTkyYjJmYjBmNTEyZTU3MmE4YjliMGUwMTRhMzdkMmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.38Hmqqo07oJNqEF43DlReyBqsvgZ5_TXBQSvKuK3iv4)
> 
> ![image](https://private-user-images.githubusercontent.com/33194443/345817137-0214f8c3-38d5-425c-a718-3312c9b2f324.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwOTM0NTAsIm5iZiI6MTcyMDA5MzE1MCwicGF0aCI6Ii8zMzE5NDQ0My8zNDU4MTcxMzctMDIxNGY4YzMtMzhkNS00MjVjLWE3MTgtMzMxMmM5YjJmMzI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA0VDExMzkxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2M2M5MjM2OWEzMzJkOTE5Y2VmMjYzOWQyZjk2NTc2OGRhZmZmYTgxMjBlMmMzNTZiOWYwYmVkOWViZGY4NzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XREPbiF495q5jWsdRX4VkeCwP-8aog97YfLl3ITSO3A)

ok it seems you already did it. So just make a test of detection and you will see that the result from the model ONNX and TFLITE are absolute not the same. And the results from TFLITE are made of random bounding boxes.

**Comment by PetiteFleurPF** [comment]
I just finished to delete the post-processing and extracted the data so:
- With this new version - I obtained that before apply any post-processing step:
- 
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/1868c879-b414-4b6b-9b8d-083db7033059)

- With the version converted:
- 
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/59ec4433-4d7d-4019-bf0c-2070b07aff62)

**Comment by PINTO0309** [comment]
![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/a7764fd3-dbd7-452a-b6fa-15b4e8c6b89f)

Thus, I still don't understand why they claim that the output values are different, even though the results of comparing all the elements are identical. If you say that you are comparing outputs with all post-processing really removed, then they must match. 

This is because the `-cotof` option compares the ONNX output with the TensorFlow output with near-perfect precision for all elements one by one.

You always seem to paste the output as an image, but that doesn't give me any of the information I need. Does the shape of the output tensor in ONNX exactly match the shape of the output tensor in TensorFlow? If the channel positions do not match exactly, there is absolutely no point in comparing element-by-element values.

**Comment by PetiteFleurPF** [comment]
- TFLITE shape:
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/50893bbe-d9bf-4bd2-ba9e-147f9b58c2d9)
- ONNX shape:
![image](https://github.com/PINTO0309/onnx2tf/assets/45994072/65414b52-4e47-4dae-9e01-884ce23e5d78)
 it's the same shape - ONNX asks just for one more dimension but I don't think it changes anything.

When I added my post-processing steps without NMS: with ONNX - the bounding boxes made sense in terms of localization but for TFLITE no. I painted in white all the original pictures in full white to illustrate what I said (the color is function of the confidence score attached to the bounding box):
ONNX
![4_00070_00280](https://github.com/PINTO0309/onnx2tf/assets/45994072/439e60da-d1ea-4efc-9b67-23227ba42358)
TFLITE
![4_00070_00280](https://github.com/PINTO0309/onnx2tf/assets/45994072/19a25a7e-92e9-41ba-b5fe-a48c912ea0db)

**Comment by PINTO0309** [comment]
Organize. I still don't understand the claim, so this will be my last reply.

1. The logic of ONNX and the preprocessing of the TFLite inference process are exactly the same.
2. No quantization is performed.
3. Normalization must not be performed in preprocessing.
4. Whenever all input data are set to 1, the inference results always match.  (12,936 All elements match.)
5. The output of `/head/regression_head/Concat_12_output_0` and `/Softmax_output_0` all match at the element level.
6. Your final bounding box result is meaningless.
7. I'm not interested in the post-processing you wrote.
8. This is not a problem with the model generated by onnx2tf. It's a difference problem in your pre-processing or post-processing.

```
INFO: onnx_output_name: /head/regression_head/Concat_12_output_0 tf_output_name: tf.concat/concat:0 shape: (1, 3234, 4) dtype: float32 validate_result:  Matches 
INFO: onnx_output_name: /head/classification_head/Concat_12_output_0 tf_output_name: tf.concat_1/concat:0 shape: (1, 3234, 91) dtype: float32 validate_result:  Matches 
INFO: onnx_output_name: /Softmax_output_0 tf_output_name: tf.nn.softmax//Softmax:0 shape: (1, 3234, 91) dtype: float32 validate_result:  Matches 
```

**Comment by PetiteFleurPF** [comment]
The solution was "Whenever all input data are set to 1". With other TFLITE model that we used, it wasn't the case. I tested and all works perfectly now. Many many .... thanks ! :D
